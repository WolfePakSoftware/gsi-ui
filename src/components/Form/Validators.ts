import validator from 'validator';

const required = (val: any) =>
  (!val || (typeof val !== 'number' && !val.toString().length)) &&
  'This field is required.';

const email = (value: string) =>
  !validator.isEmail(value) && `${value} is not a valid email.`;

const lt = (maxLength: number) => (value: string) =>
  value.toString().trim().length > maxLength &&
  `The value must me have less than ${maxLength} symbols.`;

const gt = (minLength: number) => (value: string) =>
  value.toString().trim().length < minLength &&
  `The value must me have great than ${minLength} symbols.`;

const exact = (length: number) => (value: string) =>
  value.toString().trim().length !== length &&
  `The value must be ${length} symbols.`;

const ltNumber = (min: number) => (value: string) =>
  (!validator.isNumeric(min.toString()) ||
    !validator.isNumeric(value) ||
    Number(value) > min) &&
  `The value must be less than ${min}.`;

const gtNumber = (max: number) => (value: string) =>
  (!validator.isNumeric(max.toString()) ||
    !validator.isNumeric(value) ||
    Number(value) < max) &&
  `The value must be great than ${max}.`;

const isLatLong = () => (value: string) =>
  !validator.isLatLong(value.toString()) &&
  `The value ${value} is not a valid Longitude / Latitude.`;

const isCreditCard = () => (value: string) =>
  !validator.isCreditCard(value) &&
  `The value ${value} is not a valid Credit Card.`;

const isMACAddress = () => (value: string) =>
  !validator.isMACAddress(value) &&
  `The value ${value} is not a valid MAC Address.`;

const isURL = () => (value: string) =>
  !validator.isURL(value) && `The value ${value} is not a valid URL`;

const isPhone = () => (value: string) =>
  !validator.isMobilePhone(value, 'en-US') &&
  `The value ${value} is not a valid Phone`;

const isPostalCode = () => (value: string) =>
  !validator.isPostalCode(value, 'US') &&
  `The value ${value} is not a valid Postal Code`;

const isLatitude = (value: any) => {
  return (
    value &&
    !(
      Number.isFinite(Number(value)) &&
      Number(value) >= -90 &&
      Number(value) <= 90
    ) &&
    `The value isn't a valid Latitude.`
  );
};

export const requiredIf = (func: () => boolean) => (value: any) => {
  return func() && Validators.required(value);
};

const isLongitude = (value: any) => {
  return (
    value &&
    !(
      Number.isFinite(Number(value)) &&
      Number(value) >= -180 &&
      Number(value) <= 180
    ) &&
    `The value isn't a valid Longitude.`
  );
};

export const Validators = {
  exact,
  lt,
  email,
  required,
  gt,
  ltNumber,
  gtNumber,
  isLatLong,
  isCreditCard,
  isMACAddress,
  isURL,
  isPhone,
  isPostalCode,
  isLatitude,
  isLongitude,
  requiredIf
};
