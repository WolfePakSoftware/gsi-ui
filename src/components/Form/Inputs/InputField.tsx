import { observer } from 'mobx-react';
import * as React from 'react';
/** Blueprint */
import { IconName, InputGroup, Intent } from '@blueprintjs/core';
/** FieldState */
import { StyledInput } from './style';
import { IFieldProps } from './IFieldProps';
import { FormFieldContainer } from './FormFieldContainer';
import { Validators } from '../Validators';
import { computed } from 'mobx';
import { debounce } from 'lodash';

/**
 * Field component. Must be an observer.
 */

export interface IInputFieldProps extends IFieldProps {
  leftIcon?: IconName;
  rightElement?: JSX.Element;
  round?: boolean;
  fill?: boolean;
  tipLabel?: string;
  upperCaseFormat?: boolean;
}

export interface IState {
  start: number;
  end: number;
}

@observer
export class VInputField extends React.Component<IInputFieldProps, IState> {
  inputRef: any;
  constructor(props: IInputFieldProps) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.state = {
      start: 1,
      end: 1
    };
    this.inputRef = React.createRef();
  }

  public render() {
    const {
      label,
      labelInfo,
      fieldState,
      leftIcon,
      size,
      disabled,
      inline,
      type,
      placeholder,
      rightElement,
      round,
      id,
      className,
      layer,
      fill,
      noLabel,
      required,
      validators,
      margin,
      value,
      tipLabel,
      upperCaseFormat,
      displayRequired,
      tooltip
    } = this.props;
    let rightEl;
    if (!rightElement) {
      rightEl = <div style={{ paddingRight: 10 }} />;
    }
    if (fieldState) {
      if (required) {
        if (validators && validators.length > 0) {
          fieldState.validators(Validators.required, ...validators);
        } else {
          fieldState.validators(Validators.required);
        }
      } else if (validators && validators.length > 0) {
        fieldState.validators(...validators);
      }
    }

    return (
      <StyledInput
        className={className}
        disabled={disabled}
        inline={inline}
        intent={fieldState && fieldState.hasError ? Intent.DANGER : Intent.NONE}
        labelFor={id}
        labelInfo={labelInfo}
        layer={layer}
        fill={fill}
        noLabel={noLabel}
        margin={margin}
      >
        <FormFieldContainer
          required={required || displayRequired}
          noLabel={noLabel}
          label={label}
          fieldState={fieldState}
          tooltip={tooltip}
        >
          {tipLabel && <span className={'tipLabel'}>{tipLabel}</span>}
          <InputGroup
            large={size === 'large'}
            small={size === 'small'}
            rightElement={rightElement || rightEl}
            name={id}
            inputRef={this.inputRef}
            {...{
              round,
              leftIcon,
              type,
              disabled,
              placeholder,
              id
            }}
            onChange={this.onChange}
            value={this.valueField}
            intent={
              fieldState && fieldState.hasError ? Intent.DANGER : Intent.NONE
            }
            style={{ paddingRight: 10, textTransform: 'uppercase' }}
          />
        </FormFieldContainer>
      </StyledInput>
    );
  }

  @computed
  get valueField() {
    if (this.props.fieldState) {
      return this.props.fieldState.value || '';
    }
    if (this.props.value) {
      return this.props.value;
    }
    return '';
  }

  onChange = (e: any) => {
    if (this.props.upperCaseFormat) {
      this.setState({
        start: e.target.selectionStart,
        end: e.target.selectionEnd
      });
    }

    const parsedValue =
      (this.props.upperCaseFormat && e.target.value.toString().toUpperCase()) ||
      e.target.value;
    if (this.props.fieldState) {
      this.props.fieldState.onChange(parsedValue);
    }
    if (this.props.onChange) {
      this.props.onChange(parsedValue);
    }
    if (this.props.upperCaseFormat) {
      setTimeout(() => {
        this.inputRef.current.setSelectionRange(
          this.state.start,
          this.state.end
        );
      }, 30);
    }
  };
}
