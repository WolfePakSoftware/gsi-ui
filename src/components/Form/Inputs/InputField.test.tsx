import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import { FieldState, FormState } from 'formstate';

import { required } from '../Validators';
import { VInputField } from './InputField';
import { InputGroup } from '@blueprintjs/core';
import ReactDOM from 'react-dom';

it('renders without crashing Progress Label', () => {
  const div = document.createElement('div');
  const form = new FormState<any>({
    username: new FieldState('').validators(required)
  });
  const asd = (
    <VInputField
      fieldState={form.$.username}
      inline={true}
      label={'Username'}
      placeholder={'Username'}
      size={'large'}
      labelInfo={'(info)'}
      id="test-input"
    />
  );
  ReactDOM.render(asd, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders without crashing InputField All Properties', () => {
  const form = new FormState<any>({
    username: new FieldState('').validators(required)
  });
  const InputComponent = renderer
    .create(
      <VInputField
        fieldState={form.$.username}
        inline={true}
        label={'Username'}
        placeholder={'Username'}
        size={'large'}
        labelInfo={'(info)'}
        id="test-input"
      />
    )
    .toJSON();
  expect(InputComponent).toMatchSnapshot();
});

describe('Testing Input Field', () => {
  it('check value change in Input Field', () => {
    const form = new FormState<any>({
      username: new FieldState('').validators(required)
    });
    const wrapper = mount(
      <VInputField id="test-input" fieldState={form.$.username} />
    );
    const input = wrapper.find('input');

    input.simulate('focus');
    input.simulate('change', { target: { value: 'Changed' } });

    const inputBlue = wrapper.find(InputGroup);

    inputBlue.simulate('keyDown', {
      which: 27,
      target: {
        blur() {
          input.simulate('blur');
        }
      }
    });

    const value = input.render().attr('value');
    expect(value).toBe('Changed');
    // expect(form.$.username.$).toBe('Changed');
  });
});
