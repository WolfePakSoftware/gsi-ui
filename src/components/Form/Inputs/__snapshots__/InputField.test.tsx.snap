// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing InputField All Properties 1`] = `
<div
  className="bp3-form-group bp3-inline sc-bdVaJa sc-bwzfXH iEiOly"
>
  <div
    className="bp3-form-content"
  >
    <label
      className="field-label"
      title="Username"
    >
      Username
    </label>
    <div
      className="gsi-form-field-container"
    >
      <div
        className="gsi-input-and-error-container"
      >
        <div
          className="bp3-input-group bp3-large"
        >
          <input
            className="bp3-input"
            id="test-input"
            name="test-input"
            onChange={[Function]}
            placeholder="Username"
            style={
              Object {
                "paddingRight": 10,
                "textTransform": "none",
              }
            }
            value=""
          />
          <span
            className="bp3-input-action"
          >
            <div
              style={
                Object {
                  "paddingRight": 10,
                }
              }
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
